First pass at creating a client interface ClientIface

jglukasik opened the pull request:
This PR addresses https://github.com/ipfs/ipfs-cluster/issues/447

I'm relatively new to Go and it's concept of interfaces, so if I'm doing this all wrong, don't hesitate to tell me so :slightly_smiling_face: 

In creating the new interface at `api/rest/client_interface/client_interface.go`, I had to move the definition of the `StatusFilterParams` type out of `api/rest/client/methods.go`, and into `api/types.go`, so that it could be used both in the new interface, and in the current client implementation.

I wasn't sure where to *use* the new interface. I looked for places that `c *Client` is being used that could instead use the interface. I ended up changing the tests in `methods_test.go`, does that sound right to you? Is there anywhere else I should be using this interface?

`make test` passes, and `make all` builds the binaries fine. Manually testing the newly built binaries, everything seems fine. I was able to bootstrap to an existing ipfs-cluster-service I have running on a VPS, and I was able to confirm that `ipfs-cluster-ctl add ~/testfile`  works, and it is successfully replicated to the other peer (from checking `ipfs-cluster-ctl status`)

---
(I also made a small change to  `package.json` to add a minimum `goversion`. This fixes an issue I had where i was trying to build with an older version of golang and was getting an error:
```
../../../gx/ipfs/QmYmsdtJ3HsodkePE3eU3TsCaP2YvPZJ4LoXnNkDE5Tpt7/go-multiaddr/protocols.go:6:2: cannot find package "math/bits" in any of:
        /usr/lib/go-1.6/src/math/bits (from $GOROOT)
        /home/jgl/go/src/math/bits (from $GOPATH)
```
I found the fix for this problem here https://github.com/sahib/brig/issues/13

If you would prefer to keep this change outside of this specific PR, lmk and I can create a new one)

GitCop commented:
There were the following issues with your Pull Request

 * Commit: fc4342edc4aaa2ed4b97638ca5bc22d7d49ea12b
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

Code contribution guidelines for IPFS Cluster are available at https://cluster.ipfs.io/developer/contribute/#code-contribution-guidelines

---
This message was auto-generated by https://gitcop.com

lanzafame commented on api/rest/client_interface/client_interface.go:
```diff
@@ -0,0 +1,43 @@
+	peer "github.com/libp2p/go-libp2p-peer"
+)
+
+type ClientIface interface {
```
If we move this file to the `rest` package, I would rename this to just `Client`. This will be inline with the naming of our other interfaces, for examples see the ipfscluster file: https://github.com/ipfs/ipfs-cluster/blob/753322cdc1e6d5459bace4fb2fd647f529a56278/ipfscluster.go#L70-L94

hsanjuan commented on api/rest/client_interface/client_interface.go:
```diff
@@ -0,0 +1,43 @@
+package client_interface
```
I tend to disagree. We shouldn't need to require to import and compile the rest API and its subdependencies to implement a client. I'm thinking of a future scenario where every module has been split to their own repos etc.

Perhaps `api/rest/client/interface.go` and `api/rest/client/<give name to current client>` as submodule? I'm not sure.

hsanjuan replied:
On second thought, perhaps we can have the client AND the interface in `api/rest/client/` but made the current client private and return it with `NewDefaultClient` or `New`, thus providing a default implementation for the interface. This would fix naming problems.

jglukasik replied:
I have no strong opinions either way, in the next commit I'll try putting both the interface and the default implementation in `api/rest/client/client.go`

@hsanjuan , how would the `NewDefaultClient` you're proposing differ from the currently defined `func NewClient(cfg *Config) (*Client, error)` on [client.go:96](https://github.com/ipfs/ipfs-cluster/blob/master/api/rest/client/client.go#L96) ?

lanzafame replied:
@jglukasik `NewDefaultClient` func signature would become `func NewClient(cfg *Config) (Client, error)` with the returned `Client` being the interface and not the concreate client type, hence no pointer.

hsanjuan commented on api/types.go:
```diff
@@ -894,3 +894,12 @@ type IPFSRepoStat struct {
+
+// StatusFilterParams contains the parameters required
+// to filter a stream of status results.
+type StatusFilterParams struct {
```
If client and interface live in the same package, this should move back to it.

hsanjuan commented on api/rest/clientInterface/clientInterface.go:
```diff
@@ -0,0 +1,45 @@
+
+	GetConnectGraph() (api.ConnectGraphSerial, error)
+	WaitFor(ctx context.Context, fp api.StatusFilterParams) (api.GlobalPinInfo, error)
+	AddMultiFile(multiFileR *files.MultiFileReader, params *api.AddParams, out chan<- *api.AddedOutput) error
```
Move up next to `Add`

hsanjuan commented on api/rest/clientInterface/clientInterface.go:
```diff
@@ -0,0 +1,45 @@
+	Version() (api.Version, error)
+
+	GetConnectGraph() (api.ConnectGraphSerial, error)
+	WaitFor(ctx context.Context, fp api.StatusFilterParams) (api.GlobalPinInfo, error)
```
We should extract `WaitFor` out of the Client and have it as an utility function like:

`WaitFor(ctx context.Context, c *Client[Iface], fp StatusFilterParams) (api.GlobalPinInfo, error)`.

The code in it does not depend on a specific client implementation so this way it can be re-used by multiple ones.

hsanjuan commented on package.json:
```diff
@@ -4,7 +4,8 @@
   "gx": {
-    "dvcsimport": "github.com/ipfs/ipfs-cluster"
+    "dvcsimport": "github.com/ipfs/ipfs-cluster",
+    "goversion": "1.10"
```
good :+1:

hsanjuan commented on api/rest/clientInterface/clientInterface.go:
```diff
@@ -0,0 +1,45 @@
+
+// ClientIface defines the interface to be used by API clients to interact with
+// the ipfs-cluster-service
+type ClientIface interface {
```
We will need to document each of these methods here, if we make the client private. Basically consists on copying the docstrings for each method in the current client to the methods here.

GitCop commented:
There were the following issues with your Pull Request

 * Commit: f344dd660bdbacbd3d0d3e65e6c737bbeea4eb06
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

 * Commit: 5e639fde169097e1b9a3c04751aa1f44f9cc0ec8
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

 * Commit: fa25f49a9c8ced98012ae0d08b3c8bc7365eeb5f
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

Code contribution guidelines for IPFS Cluster are available at https://cluster.ipfs.io/developer/contribute/#code-contribution-guidelines

---
This message was auto-generated by https://gitcop.com

GitCop commented:
There were the following issues with your Pull Request

 * Commit: f344dd660bdbacbd3d0d3e65e6c737bbeea4eb06
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

 * Commit: 5e639fde169097e1b9a3c04751aa1f44f9cc0ec8
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

 * Commit: 47ebd9a37f4515d54ce3939840d79df47465b9af
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

Code contribution guidelines for IPFS Cluster are available at https://cluster.ipfs.io/developer/contribute/#code-contribution-guidelines

---
This message was auto-generated by https://gitcop.com

GitCop commented:
There were the following issues with your Pull Request

 * Commit: f344dd660bdbacbd3d0d3e65e6c737bbeea4eb06
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

 * Commit: 5e639fde169097e1b9a3c04751aa1f44f9cc0ec8
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

 * Commit: 47ebd9a37f4515d54ce3939840d79df47465b9af
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

Code contribution guidelines for IPFS Cluster are available at https://cluster.ipfs.io/developer/contribute/#code-contribution-guidelines

---
This message was auto-generated by https://gitcop.com

jglukasik commented:
Apologies for the delay on this :sweat:

I think I addressed all the feedback (moved the interface into `client.go`, renamed it `Client`, changed the `WaitFor` helper, added comments), but let me know if i missed anything.

I also ran into a small issue testing. Running `make test` at the root of the `ipfs-cluster` repo was not running the tests in `api/rest/client/client_test.go`. From some [cursory Googling](https://stackoverflow.com/questions/16353016/how-to-go-test-all-testings-in-my-project), the line in the Makefile `go test -loglevel "CRITICAL" -v ./...` seems like it should be running all the tests, but I haven't investigated further to find out why.  I just did `cd api/rest/client/ && go test` as a workaround.

lanzafame commented:
> Apologies for the delay on this 

No worries, thanks for coming back to it :smile: 

> I also ran into a small issue testing. Running `make test` at the root of the `ipfs-cluster` repo was not running the tests in `api/rest/client/client_test.go`. From some [cursory Googling](https://stackoverflow.com/questions/16353016/how-to-go-test-all-testings-in-my-project), the line in the Makefile `go test -loglevel "CRITICAL" -v ./...` seems like it should be running all the tests, but I haven't investigated further to find out why. I just did `cd api/rest/client/ && go test` as a workaround.

I have also hit this issue, I think it may be because not all tests accept the `loglevel` flag, which is a cluster-specific flag, not a `go test` flag. I just run `go test -v ./...` to hit all tests. 

> I forgot to sign my commits until I pushed them here. I tried cherry-picking the commits to a new branch and signing them there, then merging and pushing that branch, but it looks like that wasn't enough to please the GitCop  Any tips on how to resolve those?

The easiest way would be to rebase all the commits and edit the commit comments to add the license footer. Probably worth squashing some of the commits while doing that as well.

hsanjuan commented on api/rest/client/client.go:
```diff
@@ -79,9 +149,9 @@ type Config struct {
-type Client struct {
+// DefaultClient provides methods to interact with the ipfs-cluster API. Use
+// NewDefaultClient() to create one.
+type DefaultClient struct {
```
you can leave this unexported as `defaultClient`.

hsanjuan commented on api/rest/client/client.go:
```diff
@@ -92,10 +162,10 @@ type Client struct {
-// NewClient initializes a client given a Config.
-func NewClient(cfg *Config) (*Client, error) {
+// NewDefaultClient initializes a client given a Config.
+func NewDefaultClient(cfg *Config) (*DefaultClient, error) {
```
and here you should be able to return `Client` (if `defaultClient` correctly implements the interface)

jglukasik commented:
Thank you for the help @hsanjuan , just pushed a clean commit.

After changing the function signature of `NewDefaultClient`, i had to make some changes to the tests in `client_test.go`. There are tests that are testing behavior specific to the `defaultClient` implementation (like `hostname` or `config`), so I had to explicitly cast the Client interface as this specific implementation with `dc := c.(*defaultClient)`. I *think* this is ok and fits the original intent of the tests, but if you'd like to do something different there let me know.

hsanjuan commented:
> . I think this is ok

Agreed
